File: java/socketd/src/main/java/org/noear/socketd/transport/core/ChannelBase.java
package org.noear.socketd.transport.core;

import org.noear.socketd.transport.core.internal.Frames;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 通道基类
 *
 * @author noear
 * @since 2.0
 */
public abstract class ChannelBase implements Channel {
    //最大请求数（根据请求、响应加减计数）
    private final Config config;

    private final Map<String, Object> attachments = new ConcurrentHashMap<>();
    private HandshakeInternal handshake;
    //是否已关闭（用于做关闭异常提醒）//可能协议关；可能用户关
    private int isClosed;


    public ChannelBase(Config config) {
        this.config = config;
    }


    @Override
    public Config getConfig() {
        return config;
    }

    @Override
    public <T> T getAttachment(String name) {
        return (T) attachments.get(name);
    }

    @Override
    public void setAttachment(String name, Object val) {
        if (val == null) {
            attachments.remove(name);
        } else {
            attachments.put(name, val);
        }
    }


    @Override
    public int isClosed() {
        return isClosed;
    }

    @Override
    public void close(int code) {
        isClosed = code;
        attachments.clear();
    }

    @Override
    public void setHandshake(HandshakeInternal handshake) {
        this.handshake = handshake;
    }


    @Override
    public HandshakeInternal getHandshake() {
        return handshake;
    }

    @Override
    public void sendConnect(String uri) throws IOException {
        send(Frames.connectFrame(getConfig().getIdGenerator().generate(), uri), null);
    }

    @Override
    public void sendConnack(Message connectMessage) throws IOException {
        send(Frames.connackFrame(connectMessage), null);
    }

    @Override
    public void sendPing() throws IOException {
        send(Frames.pingFrame(), null);
    }

    @Override
    public void sendPong() throws IOException {
        send(Frames.pongFrame(), null);
    }

    @Override
    public void sendClose() throws IOException {
        send(Frames.closeFrame(), null);
    }

    @Override
    public void sendAlarm(Message from, String alarm) throws IOException {
        send(Frames.alarmFrame(from, alarm), null);
    }
}

File: python/socket_websocket/socketd/core/ChannelBase.py
import time
from abc import ABC

from socketd.core.Channel import Channel
from socketd.core.Frames import Frames


class ChannelBase(Channel, ABC):
    def __init__(self, config):
        self.config = config
        self.requests = 0
        self.handshake = None
        self.live_time = 0
        self.attachments = {}
        self._is_closed = False

    def get_attachment(self, name):
        return self.attachments.get(name, None)

    def set_attachment(self, name, val):
        self.attachments[name] = val

    def is_closed(self):
        return self._is_closed

    def get_requests(self):
        return self.requests

    def set_handshake(self, handshake):
        self.handshake = handshake

    def get_handshake(self):
        return self.handshake

    def set_live_time(self):
        self.live_time = int(time.time() * 1000)

    def get_live_time(self):
        return self.live_time

    async def send_connect(self, uri):
        await self.send(Frames.connectFrame(self.config.get_id_generator()().__str__(), uri), None)

    async def send_connack(self, connect_message):
        await self.send(Frames.connackFrame(connect_message), None)

    async def send_ping(self):
        await self.send(Frames.pingFrame(), None)

    async def send_pong(self):
        await self.send(Frames.pongFrame(), None)

    async def send_close(self):
        await self.send(Frames.closeFrame(), None)

    async def close(self, code: int = 1000,
                    reason: str = "", ):
        self._is_closed = True
        self.attachments.clear()

    def get_config(self) -> 'Config':
        return self.config

